crates/conformance/src/outbound_pg.rs

```diff
@@ -0,0 +1,162 @@
+    /// parameters.
+    ///
+    /// The statement parameters are serialized as "<type>:<value>" pairs, e.g. "boolean:true", "uint8:42",
+    /// "str:hello", etc.
+    pub execute: Result<(), String>,
+
+    /// Result of the PostgreSQL query execution test
+    ///
+    /// This test calls into the module using the requested [`super::InvocationStyle`] with three or more arguments
+    /// in the form "outbound-pg-query <address> <query> [<parameter>...]", where "<addresss>", "<query>", and zero
+    /// or more "<parameter>" values are the parameters to be passed to outbound-pg.wit's `query` function.  The
+    /// test will assert that said function is called exactly once with the expected parameters.
+    ///
+    /// The query parameters are serialized as "<type>:<value>" pairs, e.g. "boolean:true", "uint8:42",
+    /// "str:hello", etc.
+    pub query: Result<(), String>,
+}
+
+wit_bindgen_wasmtime::export!("../../wit/ephemeral/outbound-pg.wit");
+
+#[derive(Default)]
+pub(super) struct OutboundPg {
+    execute_map: HashMap<(String, String, String), u64>,
+    query_map: HashMap<(String, String, String), RowSet>,
+}
+
+impl outbound_pg::OutboundPg for OutboundPg {
+    fn execute(
+        &mut self,
+        address: &str,
+        statement: &str,
+        params: Vec<ParameterValue<'_>>,
+    ) -> Result<u64, PgError> {
+        self.execute_map
+            .remove(&(
+                address.to_owned(),
+                statement.to_owned(),
+                format!("{params:?}"),
+            ))
+            .ok_or_else(|| {
+                PgError::OtherError(format!(
+                    "expected {:?}, got {:?}",
+                    self.execute_map.keys(),
+                    iter::once(&(
+                        address.to_owned(),
+                        statement.to_owned(),
+                        format!("{params:?}")
+                    ))
+                ))
+            })
+    }
+
+    fn query(
+        &mut self,
+        address: &str,
+        statement: &str,
+        params: Vec<ParameterValue<'_>>,
+    ) -> Result<RowSet, PgError> {
+        self.query_map
+            .remove(&(
+                address.to_owned(),
+                statement.to_owned(),
+                format!("{params:?}"),
+            ))
+            .ok_or_else(|| {
+                PgError::OtherError(format!(
+                    "expected {:?}, got {:?}",
+                    self.query_map.keys(),
+                    iter::once(&(
+                        address.to_owned(),
+                        statement.to_owned(),
+                        format!("{params:?}")
+                    ))
+                ))
+            })
+    }
+}
+
+pub(super) fn test(store: &mut Store<Context>, pre: &InstancePre<Context>) -> Result<PgReport> {
+    Ok(PgReport {
+        execute: {
+            store.data_mut().outbound_pg.execute_map.insert(
```

Comment:
suggestion: break these out into `test_execute` and `test_query` methods to reduce nesting and make the boundary pop out more